### PR TITLE
feat(@angular/ssr): dynamic route resolution using Angular router

### PR DESCRIPTION
**Pull request by alan-agius4**
This enhancement eliminates the dependency on file extensions for server-side rendering (SSR) route handling, leveraging Angular's router configuration for more dynamic and flexible route determination. Additionally, configured redirectTo routes now correctly respond with a 302 redirect status.

The new router uses a radix tree for storing routes. This data structure allows for efficient prefix-based lookups and insertions, which is particularly crucial when dealing with nested and parameterized routes.

This change also lays the groundwork for potential future server-side routing configurations, further enhancing the capabilities of Angular's SSR functionality.

### CODE REVIEW

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+ * Initializes an Angular platform, bootstraps the application or module,
+ * and retrieves routes from the Angular router configuration. Handles both module-based and
+ * function-based bootstrapping. Yields the resulting routes as `RouteResult` objects.
+ *
+ * @param bootstrap - A function that returns a promise resolving to an `ApplicationRef` or an Angular module to bootstrap.
+ * @param document - The initial HTML document used for server-side rendering.
+ * @param url - The URL for server-side rendering.
+ * @returns An async iterator yielding `RouteResult` objects.
+ */
+export async function* getRoutesFromAngularRouterConfig(
```
Comment (AndrewKushnir) [inline packages/angular/ssr/src/routes/ng-routes.ts:62]: Do we use a similar logic when extracting routes for SSG purposes? If that's the case, I'm wondering if we can eventually share the logic to avoid duplication and simplify maintenance.

File: packages/angular/ssr/src/utils/url.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    if (part[0] === '/') {
+      start = 1;
+    }
+
+    let end: -1 | undefined;
+    if (part[part.length - 1] === '/') {
+      end = -1;
+    }
+
+    const normalizePart = start || end ? part.slice(start, end) : part;
```
Comment (AndrewKushnir) [inline packages/angular/ssr/src/utils/url.ts:61]: nit:
```suggestion
    const normalizePart = start ?? end ? part.slice(start, end) : part;
```

**Review by AndrewKushnir — APPROVED** [review]
@alan-agius4 looks great, thanks 👍

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+    return bestMatchMetadata;
+  }
+
+  /**
+   * Converts the route tree into a nested object representation.
+   * @returns A nested object where each key is a route segment and each value is an object containing
+   *          metadata and possibly further nested route segments.
+   */
+  toObject(): RouteNodeObjectRepresentation {
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:164]: Question: What is the use case for `toObject` and `fromObject`? They seem to only be used in tests.

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  /** The route segment for this node. */
+  route: string;
+}
+
+/**
+ * A node in a route tree.
+ */
+interface RouteTreeNode {
+  /** The value associated with this node (a route segment or wildcard). */
+  segment: string;
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:38]: Question: How is `segment` different from `metadata.route`?

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+      node: RouteTreeNode,
+      currentPath: string,
+      currentObj: RouteNodeObjectRepresentation,
+    ) => {
+      if (node.metadata) {
+        currentObj.metadata = node.metadata;
+      }
+
+      if (node.children) {
+        for (const [segment, childNode] of node.children) {
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:180]: TIL: I always thought you needed `node.children.entries()`, I didn't realize `Map` was directly iterable like this!

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+   */
+  toObject(): RouteNodeObjectRepresentation {
+    const result: RouteNodeObjectRepresentation = {};
+
+    /**
+     * Recursively traverses the route tree to build the nested object.
+     */
+    const traverse = (
+      node: RouteTreeNode,
+      currentPath: string,
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:172]: Nit: Is this used for anything? Looks like we build up the path string recursively but don't do anything with it.

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+  static fromObject(nestedObject: RouteNodeObjectRepresentation): RouteTree {
+    const tree = new RouteTree();
+
+    /**
+     * Recursively traverses the nested object to insert routes into the tree.
+     */
+    const traverse = (currentObj: RouteNodeObjectRepresentation, currentPath: string) => {
+      if (currentObj.metadata) {
+        // Insert the current path with its metadata into the route tree
+        tree.insert(currentPath, currentObj.metadata);
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:208]: Question: How does this work with ordering? `tree.insert` is sensitive to ordering right? Is `RouteTree.fromObject(tree.toObject())` guaranteed to preserve ordering and have the same match behavior?

Even if that does work, wouldn't this imply that `RouteTree.fromObject` is sensitive to the ordering of properties in `children: { /* ... */ }`? That feels brittle to me and easy to not understand when using this API.

File: packages/angular/ssr/src/routes/router.ts [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  requiresRoutesToBeBuilt = true;
+
+  /**
+   * Creates an instance of the Router.
+   *
+   * @param manifest - An instance of `AngularAppManifest` that contains route information.
+   * The manifest provides routing configuration and other details necessary for route management.
+   */
+  constructor(private manifest: AngularAppManifest) {
+    if (manifest.routes) {
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/router.ts:34]: Question: Why do we have this branch? Why is it that sometimes the manifest will have routes and sometimes it won't?

I typically try to lean away from having "init" functions like `buildRoutes` which are easy to forget to call or call out of order (ex. accidentally calling `match` before `buildRoutes` is done). I'm wondering if there's a way to refactor this so the `ServerRouter` always has access to a list of routes (or at least a consistent way to get them).

File: packages/angular/ssr/src/utils/ng.ts [hunk 9]
```diff
@@ -28,13 +29,20 @@ import { renderApplication, renderModule } from '@angular/platform-server';
 export function renderAngular(
   html: string,
   bootstrap: Type<unknown> | (() => Promise<ApplicationRef>),
-  url: string,
+  url: URL,
```
Comment (dgp1130) [inline packages/angular/ssr/src/utils/ng.ts:32]: Consider: Call this `unnormalizedUrl` and then switch `normalizedUrl` -> `url`. That way the default "url" concept is the normalized version. This makes it harder for future changes to this function to accidentally use the unnormalized version.

File: packages/angular/ssr/src/utils/url.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+ * This function takes multiple URL segments, normalizes them by removing leading
+ * and trailing slashes where appropriate, and then joins them into a single URL.
+ *
+ * @param parts - The parts of the URL to join. Each part can be a string with or without slashes.
+ * @returns The joined URL string, with normalized slashes.
+ *
+ * @example
+ * ```js
+ * joinUrlParts('path/', '/to/resource'); // '/path/to/resource'
+ * joinUrlParts('/path/', 'to/resource'); // '/path/to/resource'
```
Comment (dgp1130) [inline packages/angular/ssr/src/utils/url.ts:38]: Question: Do we need to care about cases like `joinUrlParts('unnecessary////slashes')`?

File: packages/angular/ssr/src/utils/url.ts [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+ * ```
+ */
+export function joinUrlParts(...parts: string[]): string {
+  // Initialize an array with an empty string to always add a leading slash
+  const normalizeParts: string[] = [''];
+
+  for (const part of parts) {
```
Comment (dgp1130) [inline packages/angular/ssr/src/utils/url.ts:45]: Consider: I think a map and filter might be appropriate here?

```typescript
const normalizedParts = parts.map(() => { /* ... */ })
    .filter((part) => part !== '');
```

The need for the filter is kind of unfortunate and makes this less convenient than it should be, something to think about for you.

File: packages/angular/ssr/src/utils/url.ts [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+ * @param url - The URL object to process.
+ * @returns A new URL object with `/index.html` removed from the path, if it was present.
+ *
+ * @example
+ * ```typescript
+ * const originalUrl = new URL('http://www.example.com/page/index.html');
+ * const cleanedUrl = stripIndexHtmlFromURL(originalUrl);
+ * console.log(cleanedUrl.href); // Output: 'http://www.example.com/page'
+ * ```
+ */
```
Comment (dgp1130) [inline packages/angular/ssr/src/utils/url.ts:86]: I appreciate all the detailed JSDoc explanations for these functions with clear examples. It's very helpful for maintainability here!

File: packages/angular/ssr/src/utils/url.ts [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+export function stripIndexHtmlFromURL(url: URL): URL {
+  if (url.pathname.endsWith('/index.html')) {
+    const modifiedURL = new URL(url);
+    // Remove '/index.html' from the pathname
+    modifiedURL.pathname = modifiedURL.pathname.slice(0, -11);
```
Comment (dgp1130) [inline packages/angular/ssr/src/utils/url.ts:91]: Suggestion: The `-11` is kind of magical. I think taking the length of a literal would be more clear?

```typescript
modifiedURL.pathname = modifiedURL.pathname.slice(0, -'/index.html'.length);
```

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+   * @returns A nested object where each key is a route segment and each value is an object containing
+   *          metadata and possibly further nested route segments.
+   */
+  toObject(): RouteNodeObjectRepresentation {
+    const result: RouteNodeObjectRepresentation = {};
+
+    /**
+     * Recursively traverses the route tree to build the nested object.
+     */
+    const traverse = (
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:170]: Consider: Make a generic `traverse` or `walk` function which just returns a generator which iterates over all the nodes in the tree. Then `toObject` can use that to iterate and focus on building the object representation.

This would allow better decoupling between the two problems of 1) walking the `RouteTree` and 2) building the object representation.

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+   *                     metadata and possibly further nested route segments.
+   * @returns A RouteTree constructed from the nested object.
+   */
+  static fromObject(nestedObject: RouteNodeObjectRepresentation): RouteTree {
+    const tree = new RouteTree();
+
+    /**
+     * Recursively traverses the nested object to insert routes into the tree.
+     */
+    const traverse = (currentObj: RouteNodeObjectRepresentation, currentPath: string) => {
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:205]: Consider: You could also make a separate `walk` function for `RouteNodeObjectRepresentation` for better decoupling here too.

File: packages/angular/ssr/src/routes/ng-routes.ts (hunk 1 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/ng-routes.ts:62]: Yes, that is a goal.

File: packages/angular/ssr/src/assets.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+  constructor(private manifest: AngularAppManifest) {}
+
+  /**
+   * Retrieves the content of a server-side asset using its path.
+   *
+   * @param path - The path to the server asset.
+   * @returns A promise that resolves to the asset content as a string.
+   * @throws Error If the asset path is not found in the manifest, an error is thrown.
+   */
+  async getServerAsset(path: string): Promise<string> {
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/angular/ssr/src/assets.ts:29]: Server assets are limited to CSS and HTML.

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+
+    const baseHref =
+      injector.get(APP_BASE_HREF, null, { optional: true }) ??
+      injector.get(PlatformLocation).getBaseHrefFromDOM();
+
+    if (router.config.length === 0) {
+      // No routes found in the configuration.
+      yield { route: baseHref };
+    } else {
+      const compiler = injector.get(Compiler);
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/ng-routes.ts:106]: This is not the compiler package but the Compiler class in core.
https://github.com/angular/angular/blob/37bf7dee9917498f47bfd93d5d0bf9ce37f1877c/packages/core/src/linker/compiler.ts#L53

The router is on this https://github.com/angular/angular/blob/37bf7dee9917498f47bfd93d5d0bf9ce37f1877c/packages/router/src/router_config_loader.ts#L47

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+  routes: Route[],
+  compiler: Compiler,
+  parentInjector: Injector,
+  parentRoute: string,
+  baseHref: string,
+): AsyncIterableIterator<RouteResult> {
+  for (const route of routes) {
+    const { path, redirectTo, loadChildren, children } = route;
+
+    if (path === undefined) {
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/ng-routes.ts:140]: Undefined is actually threaded as an empty string. So updated the implementation.

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+      yield* traverseRoutesConfig(children, compiler, parentInjector, currentRoutePath, baseHref);
+    }
+
+    if (loadChildren) {
+      // Load and process lazy-loaded child routes.
+      const loadedChildRoutes = await loadChildrenHelper(
+        route,
+        compiler,
+        parentInjector,
+      ).toPromise();
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/ng-routes.ts:175]: Since Angular still supports version RxJs 6, we need to use `toPromise`, `firstValueFrom` is only available from RxJs version 7.

It might be worth discussing with the framework team to determine when we plan to drop support for RxJs 6, especially since it is not used in G3

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+    };
+
+    if (children?.length) {
+      // Recursively process child routes.
+      yield* traverseRoutesConfig(children, compiler, parentInjector, currentRoutePath, baseHref);
+    }
+
+    if (loadChildren) {
+      // Load and process lazy-loaded child routes.
+      const loadedChildRoutes = await loadChildrenHelper(
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/ng-routes.ts:171]: I don’t think it would make a significant difference. The `loadChildrenHelper` function operates on the same thread and does not involve HTTP calls or other asynchronous operations that would require context switching. In this case, parallelizing might actually hurt performance rather than improve it.

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+      // Parameterized route segment, replace it with a wildcard
+      const normalizedSegment = segment[0] === ':' ? '*' : segment;
+      let childNode = node.children?.get(normalizedSegment);
+
+      if (!childNode) {
+        childNode = {
+          segment: normalizedSegment,
+          insertionIndex: -1,
+        };
+        node.children ??= new Map();
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts:94]: Originally, my goal was to avoid creating potentially hundreds of redundant `Map` instances in the leaf nodes for large applications, especially when considering i18n.

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+        node.children.set(normalizedSegment, childNode);
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts:95]: In `/foo/*/bar` and `/foo/*/baz` the segments `bar` and `baz` are different:

```
/foo/*/bar    //  http://something.com/foo/category/bar
/foo/*/baz    //  http://something.com/foo/product/baz
```

On the other hand,  `/foo/:bar` is the same as `/foo/:baz` the Angular router does not currently handle this as a rejection. Instead, it simply fails to match the second route:

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+     * Recursively traverses the route tree to find the best match.
+     */
+    const traverse = (currentNode: RouteTreeNode, remainingSegments: string[]): void => {
+      const { insertionIndex, metadata, children } = currentNode;
+      if (!remainingSegments.length) {
+        // Reached a leaf node, check if it's a better match
+        if (metadata && insertionIndex < lowestIndex) {
+          bestMatchMetadata = metadata;
+          lowestIndex = insertionIndex;
+        }
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts:129]: I am a bit confused on how any of the above would return the match based on the order it was inserted.

File: packages/angular/ssr/src/routes/route-tree.ts (hunk 3 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts:164]: These will eventually be used to create, store, and restore the routes from the manifest. This approach is designed to avoid building the routes on the first request, which can be critical for performance in edge workers or serverless deployments. Without this, the application would need to be bootstrapped and rendered multiple times for each request, impacting performance.

File: packages/angular/ssr/src/routes/route-tree.ts (hunk 4 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts:38]: The `metadata.route` represents the full route up to the current node, provided it is a terminal route, while `segment` indicates the current route segment.

For example:

```json
{
  "segment": "en",
  "child": {
    "segment": "foo",
    "metadata": {
      "route": "en/foo"
    },
    "child": {
      "segment": "*",
      "metadata": {
        "route": "en/foo/:id"
      }
    }
  }
}
```

File: packages/angular/ssr/src/routes/route-tree.ts (hunk 7 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts:208]: > Question: How does ordering affect this process? Since `tree.insert` depends on ordering, does `RouteTree.fromObject(tree.toObject())` reliably maintain that order and ensure the same matching behavior?

The test results suggest that order is preserved.

> However, if this is the case, doesn't it imply that `RouteTree.fromObject` depends on the order of properties in children: { /* ... */ }? This seems fragile and could be confusing when using the API.

Yes, the function does rely on order, but isn't order also crucial in router configuration? In both cases, the sequence matters.

Would it be clearer to change the structure from:
```js
children: {
  '': {
    children: {
      first: {
        metadata: { route: '/first' },
      },
      '*': {
        metadata: { route: '/:id' },
      },
      second: {
        metadata: { route: '/second' },
      },
    },
  },
},
```

to 

```js
children: [
  {
    segment: '',
    children: [
      {
        segment: 'first',
        metadata: { route: '/first' },
      },
      {
        segment: '*',
        metadata: { route: '/:id' },
      },
      {
        segment: 'second',
        metadata: { route: '/second' },
      },
    ],
  },
]

```

Or perhaps a flattened version of the above?

```js
routes: [
  {
    path: '/something/first',
    metadata: { route: '/something/first' },
  },
  {
    path: '/something/[:id]',
    metadata: { route: '/something/[:id]' },
  },
  {
    path: '/something/second',
    metadata: { route: '/something/second' },
  },
]

```

File: packages/angular/ssr/src/utils/ng.ts (hunk 9 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/utils/ng.ts:32]: Refactored a bit as technically the input URL is normalized.

File: packages/angular/ssr/src/utils/url.ts (hunk 10 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/utils/url.ts:38]: I don't think there is every a possibility of having that.

File: packages/angular/ssr/src/routes/route-tree.ts (hunk 14 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts:170]: Good call, updated.

File: packages/angular/ssr/src/routes/route-tree.ts (hunk 15 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts:205]: Let me think about this following what we decide in https://github.com/angular/angular-cli/pull/28183#discussion_r1712408994

File: packages/angular/ssr/src/routes/router.ts (hunk 8 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/router.ts:34]: Certainly! Here's a more concise and focused rewrite:

---

During development, `manifest.routes` is `undefined` and the routing table is generated dynamically on the first request. This speeds up the edit-refresh cycle by eliminating the need for prebuilding.

In production, however, the routing table is prebuilt and included in the manifest. While dynamic generation is convenient for development, it's less efficient in production, especially in edge or serverless environments. Prebuilding ensures better performance by preventing the need to bootstrap and render the application multiple times per request.

I understand your concern, but since these are private APIs, the risk of "forgetting to call something" isn't a significant issue.

File: packages/angular/ssr/src/routes/ng-routes.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (dgp1130) [reply packages/angular/ssr/src/routes/ng-routes.ts]: I had a feeling RxJS 6 was the reason. We can ask about the current status but I suspect nothing is meaningfully changed for now.

File: packages/angular/ssr/src/routes/ng-routes.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (dgp1130) [reply packages/angular/ssr/src/routes/ng-routes.ts]: Ok, I'm fine to leave this as is until we end up with performance data which says otherwise.

File: packages/angular/ssr/src/routes/route-tree.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (dgp1130) [reply packages/angular/ssr/src/routes/route-tree.ts]: I personally wouldn't worry too much about optimizing at this stage until we have a sense of where the performance bottlenecks are, but I'll defer to you on which improvements are likely to be significant enough wins to be worth landing up front.

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 24]
```diff
@@ -0,0 +1,251 @@
...
+   * This allows for easy iteration over the tree's nodes.
+   * @returns A generator that yields the metadata of each node in the tree.
+   */
+  private *traverse(): Generator<RouteTreeNodeMetadata> {
+    function* traverseNode(node: RouteTreeNode): Generator<RouteTreeNodeMetadata> {
+      if (node.metadata) {
+        yield node.metadata;
+      }
+
+      for (const [, childNode] of node.children) {
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:200]: Consider: If you're ignoring the key anyways, maybe iterate over `node.children.values()`?

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 25]
```diff
@@ -0,0 +1,251 @@
...
+        yield* traverseNode(childNode);
+      }
+    }
+
+    yield* traverseNode(this.root);
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:205]: Consider: You could use define this function with the node as a parameter which defaults to `this.root` and avoids the extra function definition altogether.

```typescript
private *traverse(node: RouteTreeNode = this.route): Generator<RouteTreeNodeMetadata> {
  if (node.metadata) yield node.metadata;

  for (const [, childNode] of node.children) {
    yield* this.traverseNode(childNode);
  }
}
```

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+   */
+  static fromObject(value: RouteTreeObjectRepresentation): RouteTree {
+    const tree = new RouteTree();
+
+    for (const { route, ...metadata } of value) {
+      tree.insert(route, metadata);
+    }
+
+    return tree;
+  }
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:173]: Love how much simpler `traverse` and `traverseByRoute` make these functions. It's so much more intuitive to factor out that iteration logic.

File: packages/angular/ssr/src/routes/route-tree.ts (hunk 3 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts]: We need to make this serializable, as it will be stored in the manifest, which is generated during the build process. The manifest cannot include non-native import symbols.

Initializing a tree from a non-serializable format would add a bit more complexity.

File: packages/angular/ssr/src/routes/route-tree.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts]: Updated.

File: packages/angular/ssr/src/routes/route-tree.ts (hunk 4 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts:47]: Indeed, parameterized routes can get lost, so we could consider adding an additional field to store those parameters. However, using `metadata.route` is slightly more convenient since it eliminates the need to reconstruct the route path.

File: packages/angular/ssr/src/routes/router.ts (hunk 8 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/router.ts:34]: The URL is needed because Angular requires a URL to configure `ServerPlatformLocation` this is needed as otherwise API requests for relative requests would fail which would cause extraction to fail.

https://github.com/angular/angular/blob/d608b857c689d17a7ffa33bbb510301014d24a17/packages/platform-server/src/location.ts#L51

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+  /** The array of route configurations to process. */
+  routes: Route[];
+  /** The Angular compiler used to compile route modules. */
+  compiler: Compiler;
+  /** The parent injector for lazy-loaded modules. */
+  parentInjector: Injector;
+  /** The parent route path to prefix child routes. */
+  parentRoute: string;
+  /** The base URL of the application used to resolve relative paths. */
+  baseHref: string;
```
Comment (atscott) [inline packages/angular/ssr/src/routes/ng-routes.ts:73]: Is it necessary to carry the baseHref around and attach it to all the route results? Can't the base href logic just be applied to all results where you use them and the `RouteResults` are all just "relative to the app base"?

File: packages/angular/ssr/src/utils/url.ts [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+    if (part[0] === '/') {
+      start = 1;
+    }
+
+    let end: -1 | undefined;
+    if (part[part.length - 1] === '/') {
+      end = -1;
+    }
+
+    const normalizePart = (start ?? end) ? part.slice(start, end) : part;
```
Comment (atscott) [inline packages/angular/ssr/src/utils/url.ts:61]:
```suggestion
    let normalizedPart = part;
    if (part[0] === '/') {
      normalizedPart = part.substring(1);
    }
    if (part[part.length - 1] === '/') {
      normalizedPart = part.slice(0, -1);
    }
```
This was a bit hard to follow. I'd suggest just updating the normalized part as you go rather than saving start and end.

File: packages/angular/ssr/src/utils/url.ts [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+ * joinUrlParts('path/', '/to/resource'); // '/path/to/resource'
+ * joinUrlParts('/path/', 'to/resource'); // '/path/to/resource'
+ * ```
+ */
+export function joinUrlParts(...parts: string[]): string {
+  // Initialize an array with an empty string to always add a leading slash
+  const normalizeParts: string[] = [''];
+
+  for (const part of parts) {
+    if (!part) {
```
Comment (atscott) [inline packages/angular/ssr/src/utils/url.ts:46]:
```suggestion
    if (part === '') {
```

**Review by atscott — APPROVED** [review]

File: packages/angular/ssr/src/routes/ng-routes.ts (hunk 27 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/ng-routes.ts:73]: Great idea. I think it should be doubled. I'll give it a try tomorrow.

File: packages/angular/ssr/src/routes/route-tree.ts [hunk 30]
```diff
@@ -173,79 +183,122 @@ export class RouteTree {
...
+  /**
+   * Determines if the candidate node is prioritized based on its insertion index compared to the current best match node.
+   * A node with a lower insertion index is prioritized as it was defined earlier.
+   *
+   * @param currentBestMatchNode - The current best match node.
+   * @param candidateNode - The node being evaluated for higher priority based on insertion index.
+   * @returns `true` if the candidate node is prioritized (i.e., has a lower insertion index)
+   * compared to the current best match node, otherwise `false`.
+   */
+  private isPrioritizedByInsertionIndex(
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/route-tree.ts:281]: Consider: Rather than treating the current best match specially, consider treating the parameters symmetrically and just returning the best one. If you check for `null` as well you might be able to simplify the usage to just:

```typescript
bestMatchNode = firstByInsertionIndex(bestMatchNode, node.children.get(segment));
bestMatchNode = firstByInsertionIndex(bestMatchNode, node.children.get('*'));
bestMatchNode = firstByInsertionIndex(bestMatchNode, node.children.get('**'));
```

Alternatively you could write this like a compare function, sort an array of the candidates with it, and then pick the first one.

File: packages/angular/ssr/src/routes/router.ts [hunk 31]
```diff
@@ -6,86 +6,89 @@
...
 export class ServerRouter {
-  /** The route tree used for matching URLs to route metadata. */
-  private routeTree: RouteTree | undefined;
+  /**
+   * Creates an instance of the `ServerRouter`.
+   *
+   * @param routeTree - An instance of `RouteTree` that holds the routing information.
+   * The `RouteTree` is used to match request URLs to the appropriate route metadata.
+   */
+  constructor(private readonly routeTree: RouteTree) {}
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/router.ts:28]: Nit: Should the constructor be private so consumers go through `ServerRouter.from` instead?

File: packages/angular/ssr/src/routes/router.ts (hunk 8 above)
Comment (dgp1130) [reply packages/angular/ssr/src/routes/router.ts:50]: Thanks for adding `ServerRouter.from`, I think that greatly simplifies the contract for this class.

Are we just using the `url` from the first request? Do we build different routes based on different URL inputs? I imagine we don't really care about the URL at this stage, would it make more sense to just hard-code a `/` path?

File: packages/angular/ssr/src/routes/router.ts [hunk 32]
```diff
@@ -6,86 +6,89 @@
...
+    if (ServerRouter.#extractionPromise) {
+      return ServerRouter.#extractionPromise;
     }
 
     // Create and store a new promise for the build process.
     // This prevents concurrent builds by re-using the same promise.
-    this.buildPromise = (async () => {
-      this.routeTree = new RouteTree();
-
+    ServerRouter.#extractionPromise = (async () => {
```
Comment (dgp1130) [inline packages/angular/ssr/src/routes/router.ts:62]: Consider: You can drop the `if (ServerRouter.#extractionPromise)` with a nullish coalescing assignment.

```typescript
ServerRouter.#extractionPromise ??= (async () => { /* ... */})();
return ServerRouter.#extractionPromise;
```

**Review by dgp1130 — APPROVED** [review]
Thanks for landing these improvements, looks great!

File: packages/angular/ssr/src/routes/route-tree.ts (hunk 30 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/route-tree.ts:281]: Good idea

File: packages/angular/ssr/src/routes/router.ts (hunk 8 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/router.ts:50]: We are actually only using the protocol and host from the url as that is what is needed in this case.

https://github.com/angular/angular-cli/blob/396e83a454199763abd61b70960cfd976f370a04/packages/angular/ssr/src/routes/ng-routes.ts#L175
https://github.com/angular/angular-cli/blob/396e83a454199763abd61b70960cfd976f370a04/packages/angular/ssr/src/routes/ng-routes.ts#L181
